src/main/scala/scala/tools/refactoring/implementations/Rename.scala

```diff
@@ -122,21 +124,137 @@ abstract class Rename extends MultiStageRefactoring with MarkOccurrences with Tr
 
+      val moveToDirChanges = {
+        mkMoveToDirChanges(textChangesWithTrees.map(_._2), singleTreeSelection.symbol, newName)
+      }
+
       val textChanges = textChangesWithTrees.map(_._1)
 
-      textChanges ::: newSourceChanges
+      textChanges ::: renameFileChanges ::: moveToDirChanges
     }
 
     trace(s"Old name is ${singleTreeSelection.name}")
     if (singleTreeSelection.name == newName) Right(Nil)
     else Right(generateChanges)
   }
+
+  /*
+   * If we rename a package, it will typically be necessary to move its contents to a new directory.
+   * This method generates these changes for all files where the primary package (see below for exact definition)
+   * corresponds to the directory the files lies in. We are considering a package to be the primary package of a file,
+   * if all non package declarations lie beneath it, and no more deeply nested package satisfies this property. This
+   * is best understood by examples:
+   *
+   * Ex. 1:
+   *     package a.b //<-- primary
+   *
+   *     class C
+   *
+   * Ex. 2:
+   *     package a
+   *     package object b { //<-- primary
+   *       val x = 42
+   *     }
+   *
+   * Ex. 3:
+   *     package a { //<-- primary
+   *       package b {
+   *         class C
+   *       }
+   *       class C
+   *     }
+   */
+  private def mkMoveToDirChanges(changedTrees: List[Tree], selectedSymbol: Symbol, newName: String): List[MoveToDirChange] = {
+    if (!selectedSymbol.hasPackageFlag) {
+      Nil
+    } else {
+      changedTrees.flatMap { tree =>
+        val pkgSymbol = associatedPackageClassSymbol(tree)
+
+        pkgSymbol.flatMap  { pkgSymbol =>
+          index.rootOf(tree).flatMap { root =>
+            val pkgChain = packageChainReflectedInDirLayout(root)
+            val adaptedPkgChain = pkgChain.map {
+              case (dir, sym) if sym == pkgSymbol => (newName, pkgSymbol)
+              case other => other
+            }
+
+            if (adaptedPkgChain == pkgChain) {
+              None
+            } else {
+              val sourceFile = tree.pos.source.file
+              val newPath = {
+                val parent = parentPath(sourceFile, adaptedPkgChain.size + 1)
+                val prefix = if (parent.isEmpty) "" else parent + "/"
+                prefix + adaptedPkgChain.map(_._1).mkString("/")
+              }
+
+              Some(MoveToDirChange(sourceFile, newPath))
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private def moduleClassOf(symbol: Symbol): Option[Symbol] = symbol match {
+    case symbol: ModuleClassSymbol => Some(symbol)
+    case symbol: ModuleSymbol => Some(symbol.moduleClass)
+    case _ => None
+  }
+
+  private def associatedPackageClassSymbol(tree: Tree): Option[Symbol] = {
+    tree match {
+      case tree: ModuleDef if tree.symbol.isPackageObject => moduleClassOf(tree.symbol.enclosingPackage)
+      case tree if tree.symbol != null && tree.symbol.hasPackageFlag => moduleClassOf(tree.symbol)
```

Comment:
If you remove the `null` check, `RenameTest.testRenamePackage1002769v2` dies with a NP exception.